include/seastar/util/file.hh

```diff
@@ -64,20 +66,26 @@ namespace util {
 
 SEASTAR_MODULE_EXPORT_BEGIN
 template <typename Func>
-requires requires(Func func, input_stream<char>& in) {
-     { func(in) };
-}
-auto with_file_input_stream(const std::filesystem::path& path, Func func, file_open_options file_opts = {}, file_input_stream_options input_stream_opts = {}) {
+requires std::invocable<Func, input_stream<char>&>
+typename futurize<typename std::invoke_result_t<Func, input_stream<char>&>>::type with_file_input_stream(const std::filesystem::path& path, Func func, file_open_options file_opts = {}, file_input_stream_options input_stream_opts = {}) {
     static_assert(std::is_nothrow_move_constructible_v<Func>);
-    return open_file_dma(path.native(), open_flags::ro, std::move(file_opts)).then(
-            [func = std::move(func), input_stream_opts = std::move(input_stream_opts)] (file f) mutable {
-        return do_with(make_file_input_stream(std::move(f), std::move(input_stream_opts)),
-                [func = std::move(func)] (input_stream<char>& in) mutable {
-            return futurize_invoke(std::move(func), in).finally([&in] {
-                return in.close();
-            });
-        });
-    });
+    auto f = co_await open_file_dma(path.native(), open_flags::ro, std::move(file_opts));
+    input_stream<char> in;
```

Comment:
nit: It's a pity that input_stream<> is default-constructed, then move-assigned. Hopefully compiler can optimize it

Reply:
We can contain it in an optional but I'm not sure it's worth it.